ci-bench/src/bench_lib/cachegrind.rs

```diff
@@ -0,0 +1,237 @@
+use crate::bench_lib::Benchmark;
+use crate::cli::Side;
+use anyhow::Context;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::ops::Sub;
+use std::path::{Path, PathBuf};
+use std::process::{Child, Command, Stdio};
+
+/// A cachegrind-based benchmark runner
+pub struct CachegrindRunner {
+    /// The CPU architecture
+    arch: String,
+    /// The path to the ci-bench executable
+    ///
+    /// This is necessary because the cachegrind runner works by spawning child processes
+    executable: String,
+    /// The amount of instructions that are executed upon startup of the child process, before
+    /// actually running one of the benchmarks
+    ///
+    /// This count is subtracted from benchmark results, to reduce noise
+    overhead_instructions: u64,
+}
+
+impl CachegrindRunner {
+    /// Returns a new cachegrind-based benchmark runner
+    pub fn new(executable: String) -> anyhow::Result<Self> {
+        Self::ensure_cachegrind_available()?;
+        let arch = Self::get_arch()?;
+
+        // We don't care about the side here, so let's use `Server` just to choose something
+        let overhead_instructions = Self::run_bench_side(
+            &arch,
+            &executable,
+            u32::MAX,
+            Side::Server,
+            "calibration",
+            Stdio::piped(),
+            Stdio::piped(),
+        )?
+        .wait_and_get_instr_count()
+        .context("Unable to count overhead instructions")?;
+
+        Ok(CachegrindRunner {
+            arch,
+            executable,
+            overhead_instructions,
+        })
+    }
+
+    /// Runs the benchmark at the specified index and returns the instruction counts for each side
+    pub fn run_bench(
```

Comment:
Did you end up discovering? The index is necessary because it is passed to the child process (so it knows which specific benchmark it should run). Please let me know if you have any ideas to make this clearer 😉

Reply:
I did thanks, sorry I should have self-resolved this at the end of my review.